core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -113,43 +148,158 @@
+                afterCommitFailure = t;
+            }
+        }
+        if (afterCommitFailure != null){
+            throw afterCommitFailure;
         }
+    }
 
-        transactionManager.commit(transaction);
+    private static void fireBeforeCommit(ApplicationContext context,
+                                           PlatformTransactionManager transactionManager,
+                                           @Nullable TransactionStatus transaction) throws Throwable {
+        Throwable beforeCommitFailure = null;
 
-        for (AfterCommitTransactionListener listener : context.getBeansOfType(AfterCommitTransactionListener.class).values()) {
-            listener.afterCommit(transaction);
+        Collection<BeforeCommitTransactionListener> listeners = context.getBeansOfType
+            (BeforeCommitTransactionListener.class).values();
+        for (BeforeCommitTransactionListener listener : listeners) {
+            try {
+                listener.beforeCommit(transaction);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" newTransaction callback failed: "+ t);
+                beforeCommitFailure = t;
+            }
+        }
+        if (beforeCommitFailure != null){
+            throw beforeCommitFailure;
         }
     }
 
+    private static void fireNewTransaction(ApplicationContext context,
+                                          PlatformTransactionManager transactionManager,
+                                          @Nullable TransactionStatus transaction) throws Throwable {
+        Throwable newTransactionFailure = null;
+
+        Collection<NewTransactionListener> listeners = context.getBeansOfType
+            (NewTransactionListener.class).values();
+        for (NewTransactionListener listener : listeners) {
+            try {
+                listener.newTransaction(transaction);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" newTransaction callback failed: "+ t);
+                newTransactionFailure = t;
+            }
+        }
+        if (newTransactionFailure != null){
+            throw newTransactionFailure;
+        }
+    }
+
+    private static void fireAfterRollback(ApplicationContext context,
+                                    PlatformTransactionManager transactionManager,
+                                    @Nullable TransactionStatus transaction) throws Throwable {
+        Throwable afterRollbackFailure = null;
+
+        Collection<AfterRollbackTransactionListener> listeners = context.getBeansOfType
+            (AfterRollbackTransactionListener.class).values();
+        for (AfterRollbackTransactionListener listener : listeners) {
+            try {
+                listener.afterRollback(transaction);
+            }
+            catch (Throwable t){
```

Comment:
In previous code when a listener failed, seem no further processing, now all are processed and thrown the exception of the last one that fails. It can be fine, but want to confirm that is the intention.

Reply:
Yes that was my intension, trying to read the code and check the listener api contracts and make the application more predictable.

There was nothing explicit in the api about allowing listeners to halt the notification process (sometimes in core java api listeners are allowed to return "true" if they have addressed an event to prevent further processing).